hail/python/hail/nd/nd.py

```diff
@@ -426,6 +426,31 @@ def svd(nd, full_matrices=True, compute_uv=True):
     return construct_expr(ir, return_type, nd._indices, nd._aggregations)
 
 
+@typecheck(nd=expr_ndarray(), eigvals_only=bool)
+def eigh(nd, eigvals_only=False):
+    """Performs an eigenvalue decomposition of a symmetric matrix.
+
+    Parameters
+    ----------
+    nd : :class:`.NDArrayNumericExpression`
+        A 2 dimensional ndarray, shape(N, N).
+    eigvals_only: :class:`.bool`
+        If False (default), compute the eigenvectors and eigenvalues. Otherwise, only compute eigenvalues.
+
+    Returns
+    -------
+    - w: :class:`.NDArrayNumericExpression`
+        The eigenvalues, shape(N).
+    - v: :class:`.NDArrayNumericExpression`
+        The eigenvectors, shape(N, N). Only returned if eigvals_only is false.
```

Comment:
I've had to think about this a little bit. It would seem to me that from this description of the return type, one can do this:
```python3
wexpr, vexpr = hl.nd.eigh(my_array)
v = hl.eval(vexpr)
```
but really, you need to do
```
v, w = hl.eval(hl.nd.eigh(my_array))
```

I'm not exactly sure how we go about saying "returns an expression that, when evaluated, yields the eigenvalues and optionally, the eigenvectors if eigenvals_only is truthy" (which is rather clunky), but I think we should say something along those lines to be absolutely clear.

Reply:
The first thing should definitely work; I had a requiredness bug. Nice catch!

Reply:
Oh interesting - that's really quite cool